.circleci/config.yml

```
@@ -1,33 +1,139 @@
-version: 2
-jobs:
-    unit-test:
+version: 2.1
+executors:
+    ubuntu_docker:
         docker:
             # Match target OS of TLJH
             - image: ubuntu:18.04
-
         working_directory: ~/repo
 
+
+commands:
+    setup_venv:
+        description: Setup hub venv
         steps:
-            - checkout
+            - run:
+                name: Setup venv
+                command: |
+                    python3 -m venv /srv/venv
+                    echo 'export PATH=/srv/venv/bin:$PATH' >> $BASH_ENV
 
-            # Setup Python
+    install_python:
+        description: Install python3, venv, git and make
+        steps:
             - run:
                 name: install python
                 command: |
-                    apt-get update --yes && apt-get install --yes python3 python3-venv git
+                    apt-get update --yes && apt-get install --yes python3 python3-venv git make
+
+    setup_pkgs:
+        description: Setup pkgs for integration and upgrade tests
+        steps:
+            - run:
+                name: setup python3
+                command: |
+                    apk add --no-cache python3 pytest
+            - run:
+                # Need this to indentify the circle pr branch
+                name: install curl and jq
+                command: |
+                    apk add curl curl-dev
+                    apk add jq
+
+
+    build_systemd_image:
+        steps:
+            - run:
+                name: build systemd image
+                command: |
+                    .circleci/integration-test.py build-image
+
+    basic_tests:
+        parameters:
+            # Whether or not we should run update tests
+            upgrade:
+                type: string
+                default: ""
+        steps:
+            - run:
+                name: Run basic tests
+                command: |
+                    PR_INFO_URL=https://api.github.com/repos/$CIRCLE_PROJECT_USERNAME/$CIRCLE_PROJECT_REPONAME/pulls/$CIRCLE_PR_NUMBER
+                    PR_BRANCH=`(curl -s $PR_INFO_URL | jq -r ".head.ref")`
+                    if [ "$PR_BRANCH" == "null" ];then
+                        PR_BRANCH=$CIRCLE_BRANCH
+                    fi
+                    BOOTSTRAP_PIP_SPEC=git+https://github.com/$CIRCLE_PROJECT_USERNAME/the-littlest-jupyterhub.git@$PR_BRANCH
```

We shouldn't need to extract branche and repo info from the PR because the repo and commit being built should be enough:

```suggestion
BOOTSTRAP_PIP_SPEC=git+$CIRCLE_REPOSITORY_URL@$CIRCLE_SHA1
```

However, what might be even simpler is to use a *local path* to the current checkout, rather than referring to the repo. That can be another PR, though. If the current repo is mounted as a volume to the test container or already checked out there somehow, then we should be able to do

```suggestion
BOOTSTRAP_PIP_SPEC=file:///path/to/tljh-checkout
```

---

Thanks for reviewing this @minrk.
If I understood things right, before this PR, we were actually using a local path to the current checkout:
* we were copying the current checkout code into /srv/src in the test container [here](https://github.com/jupyterhub/the-littlest-jupyterhub/blob/master/.circleci/integration-test.py#L88)
* TLJH_BOOSTRAP_PIP_SPEC was set to /srv/src in [the Dockerfile](https://github.com/jupyterhub/the-littlest-jupyterhub/blob/master/integration-tests/Dockerfile#L25)

But this approach, although simpler, led to [this error](https://github.com/jupyterhub/the-littlest-jupyterhub/issues/502) not being caught by the tests. This is why I tried keep in the test container only the integration tests and the bootstrap script and nothing else.

However, the suggestion to use the CIRCLE_SHA1 is really great and will help get rid of that ugly hack to extract the branch.

Thanks!